### PR TITLE
IA-2841 mapping select multiple to boolean

### PR DESCRIPTION
**Pull request by mestachs**
If you don't think about adding calculate you can't export the submissions containing a multi select. To remove the need about thinking about that "upfront", we will allow to map multi select to boolean type data element.

TODO :

- [x] add tests for the export
- [x] preserve mapping of "multiple choices" when uploading a new version
- implement this for other mappings
   - [ ] support multiple choices for event tracker (warn stages)
   - [ ] support multiple choices for aggregate (warn category combo option)

Also removed redux and used react query and hooks
No typescript for the moment.

Related JIRA tickets : IA-2841 IA-2952 IA-2953

## Self proofreading checklist

- [ ] Did I use eslint and black formatters
- [ ] Is my code clear enough and well documented
- [ ] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Changes

- udpate the seed command to show case that (xlsform with select_multiple, mapping for the boolean)
image

- update the mapping screen to display the coverage and options below.
image

- the backend for storing has remain the same (quite generic)
- update the exporter to handle that

## How to test

1. launch the seed command (not it won't update the xlsform on an existing seeded account)
   - `docker-compose run --rm iaso manage seed_test_data --mode=seed --dhis2version=2.39`
3. find the "Community Verification Satisfaction form"
4. find a submission
5. export it 
6. check the results in dhis2 
   - take the "export id" on the instance, it's the eventId in the url : https://play.dhis2.org/2.39.4.1/dhis-web-capture/index.html#/viewEvent?viewEventId=aMulsxS3imr
   - image
   - image

## Print screen / video

Upload here print screens or videos showing the changes

## Notes

This won't work if the choice name contains a whitespace.

### CODE REVIEW

File: hat/assets/js/apps/Iaso/domains/mappings/actions.js
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/actions.js:1]: 🥳

File: hat/assets/js/apps/Iaso/domains/mappings/components/CreateMappingVersionDialogComponent.js [hunk 1]
```diff
@@ -60,7 +48,7 @@ const CreateMappingVersionDialogComponent = ({
             payload.program = dataset;
         }
 
-        createMappingRequest(payload).then(() => {
+        createMappingRequest.mutate(payload).then(() => {
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/components/CreateMappingVersionDialogComponent.js:51]: you need to use mutate async to have a promise, or pass a onsucces function in useCreateMappingMutation options to trigger the onclose when saved

File: hat/assets/js/apps/Iaso/domains/mappings/components/CreateMappingVersionDialogComponent.js [hunk 2]
```diff
@@ -34,21 +28,15 @@ const mappingTypeOptions = [
     },
 ];
 
-const CreateMappingVersionDialogComponent = ({
-    createMappingRequest,
-    fetchSources,
-    mappingSources,
-}) => {
+const CreateMappingVersionDialogComponent = () => {
+    const createMappingRequest = useCreateMappingMutation();
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/components/CreateMappingVersionDialogComponent.js:32]: const { mutateAsync } = useCreateMappingMutation();

File: hat/assets/js/apps/Iaso/domains/mappings/components/CreateMappingVersionDialogComponent.js [hunk 3]
```diff
@@ -34,21 +28,15 @@ const mappingTypeOptions = [
...
+    const dataSourcesRequest = useDataSources();
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/components/CreateMappingVersionDialogComponent.js:33]: const  { data } = useDataSources();

File: hat/assets/js/apps/Iaso/domains/mappings/details.js [hunk 4]
```diff
@@ -58,233 +54,182 @@ const fieldTypeOptions = formatMessage => [
...
+    const fetching = currentMappingVersionQuery.isFetching;
 
-        const onUnmapQuestionMapping = () => {
-            applyPartialUpdate(currentMappingVersion.id, currentQuestion.name, {
-                action: 'unmap',
-            });
-        };
+    const indexedQuestions = currentFormVersion
+        ? Descriptor.indexQuestions(currentFormVersion.descriptor)
+        : {};
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/details.js:75]: This one could use a useMemo i think

File: hat/assets/js/apps/Iaso/domains/mappings/details.js [hunk 5]
```diff
@@ -58,233 +54,182 @@ const fieldTypeOptions = formatMessage => [
...
+    const onQuestionSelected = node => {
+        if (node && node.type === 'select all that apply') {
+            return;
+        }
+        setCurrentQuestion(node);
+        redirectToReplace(baseUrls.mappingDetail, {
+            mappingVersionId: currentMappingVersion.id,
+            questionName: Descriptor.getKey(node),
+        });
+    };
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/details.js:107]: UseCallback

File: hat/assets/js/apps/Iaso/domains/mappings/details.js [hunk 6]
```diff
@@ -58,233 +54,182 @@ const fieldTypeOptions = formatMessage => [
...
-                    }}
-                />
-                {fetching && <LoadingSpinner />}
+    if (
+        currentFormVersion &&
+        params.questionName &&
+        currentQuestion == undefined
+    ) {
+        setCurrentQuestion(indexedQuestions[params.questionName]);
+    }
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/details.js:87]: Not a big fan of a conditional setting up the state on each render, maybe a useEffect is a better solution here

File: hat/assets/js/apps/Iaso/domains/mappings/details.js (hunk 5 above)
Comment (beygorghor) [reply hat/assets/js/apps/Iaso/domains/mappings/details.js:107]: same for other mutate function after

File: hat/assets/js/apps/Iaso/domains/mappings/details.js [hunk 7]
```diff
@@ -58,233 +54,182 @@ const fieldTypeOptions = formatMessage => [
...
+                              name: currentMappingVersion.form_version.form
+                                  .name,
+                              id: currentMappingVersion.form_version.version_id,
+                              type: currentMappingVersion.mapping.mapping_type,
+                          })
+                        : formatMessage(MESSAGES.loading)
+                }
+                displayBackButton
+                goBack={() => {
+                    goBack(backParams);
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/details.js:155]: i think we have a useGoBack hook now

File: hat/assets/js/apps/Iaso/domains/mappings/details.js [hunk 8]
```diff
@@ -58,233 +54,182 @@ const fieldTypeOptions = formatMessage => [
...
-                                {currentQuestion == null &&
-                                    currentMappingVersion && (
-                                        <GeneraMappingInfo
-                                            applyUpdate={applyUpdate}
-                                            currentMappingVersion={
+                                )}
+                            {currentQuestion && (
+                                <>
+                                    <QuestionInfos question={currentQuestion} />
+                                    <br />
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/details.js:187]: 😠

File: hat/assets/js/apps/Iaso/domains/mappings/index.js [hunk 9]
```diff
@@ -1,127 +1,69 @@
...
+            <TopBar title={formatMessage(MESSAGES.dhis2Mappings)} />
+            <Box className={classes.containerFullHeightNoTabPadded}>
+                <Table
+                    data={mappingVersionsResults?.data?.mapping_versions || []}
+                    pages={pages}
+                    defaultSorted={[
+                        { id: 'form_version__form__name', desc: true },
+                        { id: 'form_version__version_id', desc: true },
+                        { id: 'mapping__mapping_type', desc: true },
+                    ]}
```
Comment (beygorghor) [inline hat/assets/js/apps/Iaso/domains/mappings/index.js:47]: Maybe use a constant for this

File: docker-compose.yml [hunk 10]
```diff
@@ -1,6 +1,8 @@
 version: "3"
 services:
   iaso:
+    stdin_open: true
+    tty: true
```
Comment (beygorghor) [inline docker-compose.yml:5]: What is this for ?

**Review by beygorghor — APPROVED** [review]
One reducer less, no more class component this is great !!! Many thanks.

Just a few remarks non blocking

File: docker-compose.yml (hunk 10 above)
Comment (kemar) [reply docker-compose.yml:5]: This allows to use an interactive shell with Docker, useful to e.g. run a Python debugger.

File: hat/assets/js/apps/Iaso/domains/mappings/components/RecursiveTreeView.js [hunk 11]
```diff
@@ -31,6 +31,7 @@ const RecursiveTreeView = props => {
 
     const onNodeSelected = (event, value) => {
         const val = indexedQuestions[value];
+        console.log('node selected ', value, val);
```
Comment (kemar) [inline hat/assets/js/apps/Iaso/domains/mappings/components/RecursiveTreeView.js:34]: We could remove this line.

File: iaso/models/forms.py [hunk 12]
```diff
@@ -195,13 +195,16 @@ def _reformat_questions(questions):
     r = []
 
     for question in questions.values():
-        if question["type"] == "note":
+        print(question)
```
Comment (kemar) [inline iaso/models/forms.py:198]: We could remove this `print`.

**Review by kemar — APPROVED** [review]
LGTM

**Pull request merged**